rust/datafusion/src/datasource/csv.rs

```diff
@@ -142,3 +143,142 @@ impl TableProvider for CsvFile {
         self.statistics.clone()
     }
 }
+/// Loads CSV data from a reader
+pub struct CsvRead<R: Read> {
```

Comment:
I wonder if there is any way to reduce the duplication between `CsvRead<R>` and `CsvFile` -- as you have done for `CsvExec`. That way we can reuse the same tests for things like schema matching.

Since the `Reader` gets `Box`ed anyways for execution, I don't think there is any performance difference using something that is generic over `R` vs a `Box<R>`

Reply:
Okay. Would you mind to introduce another dependency crate `either` ?  It's more clear than introducing a new enum.

```rs
pub struct CsvFile {
    source: Either<Box<dyn Read>,String>,
   ....
}
```

Reply:
I think we are trying to keep the number of dependencies down (as we expect applications to be built on top of DataFusion) so I think we should use a specific enum rather than another crate